python/ray/data/dataset.py

```diff
@@ -652,15 +692,19 @@ def add_column(
                 ray (e.g., num_gpus=1 to request GPUs for the map tasks).
         """
 
-        def process_batch(batch):
-            batch[col] = fn(batch)
+        def process_batch(batch: "pandas.DataFrame") -> "pandas.DataFrame":
+            batch.loc[:, col] = fn(batch)
```

Comment:
Driveby: we should use the `.loc[]` API for mutating DataFrame batches in-place.